### PR TITLE
Updating markovbot.py

### PR DESCRIPTION
**Pull request by jobarber**
@timothybeal and @textpotential,

I want to start the process of engaging this project with the open source mentality. I am proposing a few changes to `markovbot.py`, but you do not need to accept them. I would be happy to walk through them with you. I see still further improvements that could be made, but I'll leave those alone for now. Most of your code is still in place (despite the highlighting in github, which suggests that _most_ lines have been replaced; that is not the case).

I am suggesting we put some of this into a `class` in order to make data persistent and to allow multiple instances of a Markovizer (using different texts). Feel free to discuss these changes here when you have the chance. Mostly, I want to initiate the process of how these conversations take place with external contributors to this space.

@jobarber

### CODE REVIEW

**Comment by jobarber** [comment]
I have added my proposed changes for `kjvbot3.py`, although I did not get to test them. Still, it should work. And you could include other possible starts that are not three tokens in length as well, if you wish.

**Comment by timothybeal** [comment]
Thanks Justin! I’ll need to change these in the Eclipse files on my laptop and in the Flask thing which should be easy. Btw I think I’ve probably changed the kjvbot3 on Eclipse in other ways too. That’s what I get for not keeping track / updated!

**Comment by jobarber** [comment]
Sounds great, @timothybeal. There is no rush on this. We can think and talk through this. Right now, I have my own branch I/we can modify. You can always update your branch with your own changes, and then we can work on a merge (although slightly more complicated!).

**Comment by timothybeal** [comment]
Could be a good way of demo'ing github editing and forks and such this Tues?

Timothy Beal
Florence Harkness Professor of Religion
Chair, Department of Religious Studies
Case Western Reserve University
timothybeal.com

**Comment by textpotential** [comment]
Wow @jobarber those are significant changes. Could be a very good introduction to classes for the group on Tuesday. You could show how you translated (can't help myself :) code from set of functions to methods of a class and discuss the advantages.

**Comment by textpotential** [comment]
As I was reading the diff file for the main module, I couldn't help but think about the distant reading conversation, even though I haven't read that piece yet. I began to wonder what kinds of skills, reading skills, are required to read code like this, particularly in diff view, and could these skills inform other kinds of reading in helpful ways? If we are interested in encouraging close reading, perhaps this would be a good way?

File: markovbot.py [hunk 1]
```diff
@@ -1,10 +1,33 @@
-'''
+"""
 Created on Mar 10, 2016
 
 @author: timothybeal and textpotential
 
-The functions used to create a markov chain utterance and tweet if desired.
-'''
+The functions used to create a markov chain utterance and tweet, if desired.
```
Comment (jobarber) [inline markovbot.py:6]: Mainly adding example usage to the documentation here.

File: markovbot.py [hunk 2]
```diff
@@ -13,63 +36,165 @@
...
+
 def nwise(iterable, n=2):
-        if len(iterable) < n:
-            return
-        iterables = tee(iterable, n)
-        for i, iter_ in enumerate(iterables):
-            for num in range(i):
-                next(iter_)
-        return zip(*iterables)
+    """nwise([1, 2, 3, 4], n=3) => [(1, 2, 3), (2, 3, 4)]"""
```
Comment (jobarber) [inline markovbot.py:41]: Fixed spacing here.

File: markovbot.py [hunk 3]
```diff
@@ -13,63 +36,165 @@
...
+    if len(iterable) < n:
+        return
+    iterables = tee(iterable, n)
+    for i, iter_ in enumerate(iterables):
+        for num in range(i):
+            next(iter_)
+    return zip(*iterables)
 
-def build_sentence(seed, sent_tokens):
```
Comment (jobarber) [inline markovbot.py:25]: Moved this function inside the class to make it a method.

File: markovbot.py [hunk 4]
```diff
@@ -13,63 +36,165 @@
...
+        distribution.
+
+    encoding : string, 'utf-8' is default.
+        The encoding will be the encoding used to open the file,
+        if input is a file path.
+
+    token_pattern : raw string
+        The pattern to use to find the tokens in the intial sentence.
+    """
+    def __init__(self, text, encoding='utf-8',
```
Comment (jobarber) [inline markovbot.py:72]: The `__init__` method is the major addition to the code. It defines some attributes and helps preserve state from call to call.

File: markovbot.py [hunk 5]
```diff
@@ -13,63 +36,165 @@
...
-    return sentence
+        sentence = ' '.join(tokens)
+        sentence = re.sub(r'\s+([.,?!])',r'\1', sentence)
 
-def markovize(word1, word2, word3, fileid, char_limit=None):   
+        return sentence
     
-    with open(fileid, encoding='utf-8') as f:
-        text = f.read()
+    def create_prob_dist(self, current_state_len=3):
```
Comment (jobarber) [inline markovbot.py:113]: Separates the creation of the probability distribution from the actual creation of a sentence. That is good for performance when the function is called multiple times.

File: markovbot.py [hunk 6]
```diff
@@ -13,63 +36,165 @@
...
+        consumer_secret : string
+            The consumer secret for the twitter api.
+    
+        access_key : string
+            The access key for the twitter api.
+
+        access_secret : string
+            The access secret for the twitter api.
+        """
+        if not self.utterance:
```
Comment (jobarber) [inline markovbot.py:193]: Added a couple possible exceptions here. Should also add another that raises an error if tweepy is not importable, but that would need to be at the top of the file in its present form.

**Comment by jobarber** [comment]
@textpotential Great idea! I added some comments above (which you were perhaps notified about). My main change was to move the three main functions under the class definition. Then I could remove some redundancies and decouple some things based on the preserved state in the class. Great question about reading. There are a lot of interesting ideas to explore there. Perhaps we can walk through this on Tuesday. I can show them the split diff view as well, which seems to work better for the changes I made here.

**Pull request merged**